### PR TITLE
Allow baking of subdivision surfaces to importer mesh.

### PR DESCRIPTION
**Pull request by fire**
Test!

### CODE REVIEW

File: src/import/gltf_quad_importer.cpp [hunk 1]
```diff
@@ -159,20 +158,33 @@ void GLTFQuadImporter::convert_importer_meshinstance_to_quad(Object *p_meshinsta
...
+			subdiv_mesh->set_data_mesh(topology_data_mesh);
+			subdiv_mesh->set_subdiv_level(subdiv_level);
+			for (int64_t blend_shape_idx = 0; blend_shape_idx < subdiv_mesh->_get_blend_shape_count(); blend_shape_idx++) {
+				subdiv_importer_mesh->add_blend_shape(subdiv_mesh->_get_blend_shape_name(blend_shape_idx));
+			}
+			for (int64_t surface_i = 0; surface_i < subdiv_mesh->_get_surface_count(); surface_i++) {
+				Ref<Material> material = subdiv_mesh->_surface_get_material(surface_i);
+				Array blend_shape_arrays;
+				for (int64_t blend_shape_i = 0; blend_shape_i < subdiv_mesh->_get_blend_shape_count(); blend_shape_i++) {
+					blend_shape_arrays.push_back(subdiv_mesh->_surface_get_blend_shape_arrays(blend_shape_i));
```
Comment (fire) [inline src/import/gltf_quad_importer.cpp]: _get_blend_shape_count is per blendshape per surface..

File: src/import/gltf_quad_importer.cpp [hunk 2]
```diff
@@ -159,20 +158,33 @@ void GLTFQuadImporter::convert_importer_meshinstance_to_quad(Object *p_meshinsta
 		subdiv_mesh_instance->set_name(mesh_instance_name);
 		p_meshinstance->queue_free();
 	} else if (import_mode == ImportMode::BAKED_SUBDIV_MESH) {
-		BakedSubdivMeshInstance3D *subdiv_mesh_instance = memnew(BakedSubdivMeshInstance3D);
+		ImporterMeshInstance3D *subdiv_mesh_instance = memnew(ImporterMeshInstance3D);
```
Comment (tefusion) [inline src/import/gltf_quad_importer.cpp]: Bad naming from me, but the BakedSubdivMesh just exists to interface topology data and work like a mesh. The actual baking import should be done at the else if (import_mode == ImportMode::ARRAY_MESH && subdiv_level > 0).

**Comment by fire** [comment]
Todo:

1. generate_lods has two parameters to be exposed
2. some .import files are changed for no reason

File: src/import/gltf_quad_importer.cpp [hunk 3]
```diff
@@ -158,26 +158,57 @@ void GLTFQuadImporter::convert_importer_meshinstance_to_quad(Object *p_meshinsta
...
+			subdiv_mesh->set_data_mesh(topology_data_mesh);
+			subdiv_mesh->set_subdiv_level(subdiv_level);
+			for (int64_t blend_shape_idx = 0; blend_shape_idx < subdiv_mesh->_get_blend_shape_count(); blend_shape_idx++) {
+				StringName shape_name = subdiv_mesh->_get_blend_shape_name(blend_shape_idx);
+				subdiv_importer_mesh->add_blend_shape(shape_name);
+			}
+			for (int64_t surface_i = 0; surface_i < subdiv_mesh->_get_surface_count(); surface_i++) {
+				Ref<Material> material = subdiv_mesh->_surface_get_material(surface_i);
+				Array blend_shape_arrays;
+				for (int64_t blend_shape_i = 0; blend_shape_i < subdiv_mesh->_get_blend_shape_count(); blend_shape_i++) {
```
Comment (tefusion) [inline src/import/gltf_quad_importer.cpp]: Blend Shapes won't work. The only way I could think of generating them is to apply them at max to the mesh and then calculate the new relative values. Unless really needed won't be implemented, BakedSubdivMesh should instead be used directly with BakedMeshInstance3D imo.

**Comment by tefusion** [comment]
Thanks for making this! I have no idea about lod generation so this is really helpful. I'll try and fix the pipeline error. I don't think the generate_lods parameters need to be exposed within the editor setting, within C++ should be enough. I personally never used them and I think more parameters will just confuse in the importer window (cause they'd only be for arraymesh and haven't figured out how to hide single options upon option changes).

**Comment by fire** [comment]
@tefusion do you know how to debug a crash?

**Comment by tefusion** [comment]
> @tefusion do you know how to debug a crash?

Comes down to when the crash happens. If you can still set a breakpoint you can debug an extension in VS Code with this launch.json
```json
{
    "version": "0.2.0",
    "configurations": [
{
            "name": "Launch Project",
            "type": "cppdbg",
            "request": "launch",
            "program": "/path/to/godot/beta/2"
            "args": [
                "--verbose",
                "--editor",
                "--path",
                "${workspaceFolder}/project"
            ],
            "stopAtEntry": false,
            "cwd": "${workspaceFolder}",
            "environment": [],
            "externalConsole": false,
            "preLaunchTask": "build debug"
        }]}
```
and tasks.json:
```json
{
    "version": "2.0.0",
    "tasks": [
        {
{
            "label": "build debug",
            "type": "shell",
            "options": {
                "cwd": "${workspaceFolder}"
            },
            "command": "scons",
            "group": "build",
            "windows": {
                "args": [
                    "target=debug",
                    "platform=windows",
                    "-j8"
                ]
            },
            "x11": {
                "args": [
                    "target=debug",
                    "platform=x11",
                    "-j8"
                ]
            },
            "problemMatcher": "$msCompile"
        }}]}
```

**Comment by fire** [comment]
I don't know why, and debugging this is very painful as I can't seem to build opensubdiv for windows.

**Comment by tefusion** [comment]
> I don't know why, and debugging this is very painful as I can't seem to build opensubdiv for windows.

I’ll open windows and see if I can get it running myself, my most common issue was/is godot cpp not being the right version (annoying to update cause quick fix commits by me which cause merge conflicts)

Edit: Running yes, debugging no. can't help you, sry

**Comment by fire** [comment]
The dream of the bake procedure procedure is that I can reuse the ImporterMesh and ImporterMesh3D pipeline for blend shapes, bone skinning and animations.

**Comment by tefusion** [comment]
> The dream of the bake procedure procedure is that I can reuse the ImporterMesh and ImporterMesh3D pipeline for blend shapes, bone skinning and animations.

Can you elaborate more what you want in the importer, I never worked with ImporterMeshInstance3D and ImporterMesh before this. My idea with the three importer options was (again sry for bad naming and other issues, I never cleaned up the importer and kept naming variables the same as the meshinstance to be able to copy paste more) 

1. SubdivMeshInstance3D: Creates per meshinstance a new subdivision mesh for cheaper skinning and blenshapes 
2. BakedSubdivMesh: Good for just saving space, will actually bake mesh at runtime
3. ArrayMesh: Bake at import to a usable MeshInstance3D with ArrayMesh

What you currently get from the BakedSubdivMesh and ArrayMesh option is an ImporterMeshInstance which won't actually show for the user? I can merge after we clear that up, cause while I will mainly use option 1, I thought option 2 would be useful too. 

Edit: Nvm you can remove the BakedSubdivMeshInstance3D.  could you revert the do no need to generate lods cause if not adding to scene directly I think you do.

**Comment by fire** [comment]
Internally all the exporters in godot engine create ImporterMesh to be stored in ImporterMesh3D. Neither ArrayMesh or MeshInstance3D exist in the importers.

The ImporterMesh3D is converted to MeshInstance3D and the ImporterMesh is converted to ArrayMesh in the importer pipeline.

**Review by tefusion — APPROVED** [review]

**Comment by tefusion** [comment]
> Internally all the exporters in godot engine create ImporterMesh to be stored in ImporterMesh3D. Neither ArrayMesh or MeshInstance3D exist in the importers.
> 
> The ImporterMesh3D is converted to MeshInstance3D and the ImporterMesh is converted to ArrayMesh in the importer pipeline.

Shouldn't it still be replaced with an actual MeshInstance3D at the end of it? I can't really do much with just an ImporterMeshInstance (as a user)

**Comment by tefusion** [comment]
Well otherwise everything worked in a quick test and the conversion to MeshInstance can be done extra so I'll merge this now.

Thanks!

**Pull request merged**